File: problems/017_regiuni/OlegSmac.java
import java.util.*;

public class olympus17 {
    private Hashtable<Integer, List<Integer>> paths;
    private int[] components;
    private int numComponents;

    private Stack<Integer> orderVertices;
    private boolean[] marked;

    public olympus17(int n) {
        this.paths = new Hashtable<>();
        this.components = new int[n + 1];
        for (int i = 1; i < n + 1; i++) {
            components[i] = -1;
        }

        orderVertices = new Stack<>();
        marked = new boolean[n + 1];
    }

    private void findOrderVertices(Hashtable<Integer, List<Integer>> p) {
        Hashtable<Integer, List<Integer>> revPaths = new Hashtable<>();
        for (int i : p.keySet()) {
            List<Integer> listI = p.get(i);
            for (int j : listI) {
                if (revPaths.containsKey(j)) {
                    revPaths.get(j).add(i);
                }
                else {
                    List<Integer> l = new ArrayList<>();
                    l.add(i);
                    revPaths.put(j, l);
                }
            }
        }
        for (int i = 1; i < marked.length; i++) {
            if (!marked[i]) {
                dfs(i, revPaths);
            }
        }
    }

    private void dfs(int v, Hashtable<Integer, List<Integer>> revPaths) {
        marked[v] = true;
        if (revPaths.get(v) != null) {
            for (int i : revPaths.get(v)) {
                if (!marked[i]) {
                    dfs(i, revPaths);
                }
            }
            orderVertices.add(v);
        }
        else {
            orderVertices.add(v);
        }
    }

    public int findStronglyConnectedComponents() {
        // записать в стек порядок рассмотрения вершин
        findOrderVertices(paths);

        numComponents = 0;
        while (!orderVertices.isEmpty()) {
            int i = orderVertices.pop();
            if (components[i] == -1) {
                dfs(i);
                numComponents++;
            }
        }
        for (int i = 1; i < components.length; i++) {
            System.out.print(components[i] + " ");
        }
        return numComponents;
    }

    private void dfs(int v) {
        components[v] = numComponents;
        if (paths.get(v) != null) {
            for (int w : paths.get(v)) {
                if (components[w] == -1) {
                    dfs(w);
                }
            }
        }
    }

    public static void main(String[] args) {
        Scanner in = new Scanner(System.in);
        int n = in.nextInt();
        olympus17 p = new olympus17(n);
        for (int i = 1; i < n + 1; i++) {
            List<Integer> list = new ArrayList<>();
            p.paths.put(i, list);
        }
        while (in.hasNext()) {
            int v = in.nextInt();
            if (v == 0) break;
            p.paths.get(v).add(in.nextInt());
        }
        System.out.println();
        System.out.println("num components = " + p.findStronglyConnectedComponents());
    }
}

File: computer_science/017_regiuni/OlegSmac.cpp
#include<vector>
#include<unordered_map>
#include<stack>
#include<iostream>
#include<fstream>

using namespace std;

unordered_map<int, vector<int>> paths;
vector<int> components;
int numComponents;
stack<int> orderVertices;
vector<bool> marked;

void dfs(int v, unordered_map<int, vector<int>>& revPaths) {
	marked[v] = true;
	if (revPaths.find(v) != revPaths.end()) {
		for (int i : revPaths[v]) {
			if (!marked[i]) {
				dfs(i, revPaths);
			}
		}
		orderVertices.push(v);
	}
	else {
		orderVertices.push(v);
	}
}

void findOrderVertices(unordered_map<int, vector<int>>& p) {
	unordered_map<int, vector<int>> revPaths;
	for (int i = 0; i < p.size(); i++) {
		vector<int> listI = p[i];
		for (int j : listI) {
			if (revPaths.find(j) != revPaths.end()) {
				revPaths[j].push_back(i);
			}
			else {
				vector<int> l;
				l.push_back(i);
				revPaths.insert({j, l});
			}
		}
	}
	for (int i = 0; i < marked.size(); i++) {
		if (!marked[i]) {
			dfs(i, revPaths);
		}
	}
}

void dfs(int v) {
	components[v] = numComponents;
	if (paths.find(v) != paths.end()) {
		for (int w : paths[v]) {
			if (components[w] == -1) {
				dfs(w);
			}
		}
	}
}

int findStronglyConnectedComponents() {
	findOrderVertices(paths);
	numComponents = 0;
	while (!orderVertices.empty()) {
		int i = orderVertices.top();
		orderVertices.pop();
		if (components[i] == -1) {
			dfs(i);
			numComponents++;
		}
	}
	return numComponents;
}

int main() {
	ifstream in("city6.txt");
	int n;
	in >> n;
    for (int i = 0; i < n; i++) {
        components.push_back(-1);
		marked.push_back(false);
    }
    for (int i = 0; i < n; i++) {
		vector<int> vec;
        paths[i] = vec;
    }
	while (true) {
		int a;
		int b;
		in >> a;
		in >> b;
		if (a == 0 || b == 0) break;
		paths[a - 1].push_back(b - 1);
	}
	findStronglyConnectedComponents();
    ofstream out("res.txt");
	out << numComponents;
	in.close();
	out.close();
    
    cout << "Program is done.\n";
}